ref/test/utilityfunctions/isEqual.f90

```diff
@@ -0,0 +1,83 @@
+module isequalmodule
+  
+ contains 
+
+  !------------------------------------------------------------
+  !       isequal 
+  !
+  !       Function which compares equality of two real*8 floats
+  !       within a given tolerance value.
+  !------------------------------------------------------------
+
+  logical function isequal(a, b, tolerance)
+    implicit none
+
+    real*8, intent(in)   :: a, b, tolerance
+    real*8               :: absA, absB, diff
+
+    absA = abs(a)
+    absB = abs(b)
+    diff = abs(a - b)
+
+    if (a == b) then
+        isequal = .true.
+    else if ( a == 0 .or. b == 0 .or. (absA + absB < TINY(a))) then
+        ! a or b is zero or both are extremely close to it, relative error is less meaningful
+        isequal = diff < (tolerance * TINY(a))
+    else    ! use relative error
+        isequal = (diff / MIN((absA + absB), HUGE(a))) < tolerance
```

Comment:
The only case where (absA + absB) > HUGE(a) is if both absA and absB were near the upper limit of floating point numbers such that their sum is greater than HUGE(a), which is infinity.  diff/infinity is 0 which is certainly less than tolerance.  isequal would be set to true.

The upper range of 8 byte IEEE-754 floating point numbers is 1.8E308.  So if a = 1.8E308 and b=1.7E308, we would be in the above case, where isequal is true.  But those numbers are not equal.

Reply:
That is correct, this case handles comparison of numbers that are very large.  Instead of using one of the numbers, or the maximum of the two numbers, to compute the relative factor, it uses the sum of the two numbers to compute that.  And if that sum would cause an overflow, it uses HUGE instead, because that's the best you can do.

I don't see a problem here.

Reply:
OK, I see what they are doing now.  Not a problem.  Clever.